dftimewolf/lib/state.py

```diff
@@ -813,3 +827,26 @@ def PublishMessage(self,
       message: The message content.
       is_error: True if the message is an error message, False otherwise."""
     self.cursesdm.EnqueueMessage(source, message, is_error)
+
+
+def FindCallerModule() -> str:
+  """Finds the caller module runtime name.
+
+  This method is used in Get/StoreContainer to find the module name that called
+  that method. Used for debug logging to track container lifecycles.
+
+  This method looks two frames up the stack for the caller information:
+  * 0: state.py:FindCallerModule (this frame when the method is called)
+  * 1: state.py:DFTimewolfState.Get/StoreContainer()
+  * 2: <module>.<method> <- The module name we want.
+
+  In unit tests, #2 is the unit test class, which has no 'name' attribute. In
+  such a scenario, fall back to the class name.
+
+  Returns:
+    A string of the calling module runtime name if present, else the class name.
+  """
+  desired_stack_frame = 2
+  caller_locals = inspect.stack()[desired_stack_frame][0].f_locals["self"]
+  return str((caller_locals.name if hasattr(caller_locals, 'name') else
+      caller_locals.__class__.__name__))
```

Comment:
To make things a bit less magical, and avoid all this stack introspection, it might be better to change the params of `StoreContainer` so that it takes in a "caller" optional arg, and then add a function at the Module level that passes in the module's current name (much like we do with `ModuleError` currently).

```python
# in Module code

def Process(self):
  [...]
  self.StoreContainer(container.Foo())

# in module.py

def StoreContainer(self, container):
  self.state.StoreContainer(container, caller=self.name)
```

 `caller` is not super elegant but maybe it will do for now?

Reply:
That crossed my mind (and definitely has benefits, like that would be where you'd put the logging and calculate the identify the module name) but I prefer the current way because it means that all usages of `self.state.Get/StoreContainer`  don't have to be updated to `self.Get/StoreContainer`. I thought that was an acceptable trade off with going via stack introspection.